tower-http/src/request_id.rs

```diff
@@ -0,0 +1,573 @@
+//! [`ServiceBuilderExt`]: crate::ServiceBuilderExt
+//! [`Uuid`]: https://crates.io/crates/uuid
+//! [`Trace`]: crate::trace::Trace
+
+// NOTE: when uuid 1.0 is shipped we can include a `MakeRequestId` that uses that
+// See https://github.com/uuid-rs/uuid/issues/113
+
+use http::{
+    header::{HeaderName, HeaderValue},
+    Request, Response,
+};
+use pin_project_lite::pin_project;
+use std::task::{Context, Poll};
+use std::{future::Future, pin::Pin};
+use tower_layer::Layer;
+use tower_service::Service;
+
+pub(crate) const X_REQUEST_ID: &str = "x-request-id";
+
+/// Trait for producing [`RequestId`]s.
+///
+/// Used by [`SetRequestId`].
+pub trait MakeRequestId {
+    /// Try and produce a [`RequestId`] from the request.
+    fn make_request_id<B>(&mut self, request: &Request<B>) -> Option<RequestId>;
+}
+
+/// An identifier for a request.
+#[derive(Debug, Clone)]
+pub struct RequestId(HeaderValue);
+
+impl RequestId {
+    /// Create a new `RequestId` from a [`HeaderValue`].
+    pub fn new(header_value: HeaderValue) -> Self {
+        Self(header_value)
+    }
+
+    /// Gets a reference to the underlying [`HeaderValue`].
+    pub fn header_value(&self) -> &HeaderValue {
+        &self.0
+    }
+
+    /// Consumes `self`, returning the underlying [`HeaderValue`].
+    pub fn into_header_value(self) -> HeaderValue {
+        self.0
+    }
+}
+
+impl From<HeaderValue> for RequestId {
+    fn from(value: HeaderValue) -> Self {
+        Self::new(value)
+    }
+}
+
+/// Set request id headers and extensions on requests.
+///
+/// This layer applies the [`SetRequestId`] middleware.
+///
+/// See the [module docs](self) and [`SetRequestId`] for more details.
+#[derive(Debug, Clone)]
+pub struct SetRequestIdLayer<M> {
+    header_name: HeaderName,
+    make_request_id: M,
+}
+
+impl<M> SetRequestIdLayer<M> {
+    /// Create a new `SetRequestIdLayer`.
+    pub fn new(header_name: HeaderName, make_request_id: M) -> Self
+    where
+        M: MakeRequestId,
+    {
+        SetRequestIdLayer {
+            header_name,
+            make_request_id,
+        }
+    }
+
+    /// Create a new `SetRequestIdLayer` that uses `x-request-id` as the header name.
+    pub fn x_request_id(make_request_id: M) -> Self
+    where
+        M: MakeRequestId,
+    {
+        SetRequestIdLayer::new(HeaderName::from_static(X_REQUEST_ID), make_request_id)
+    }
+}
+
+impl<S, M> Layer<S> for SetRequestIdLayer<M>
+where
+    M: Clone + MakeRequestId,
+{
+    type Service = SetRequestId<S, M>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        SetRequestId::new(
+            inner,
+            self.header_name.clone(),
+            self.make_request_id.clone(),
+        )
+    }
+}
+
+/// Set request id headers and extensions on requests.
+///
+/// See the [module docs](self) for an example.
+///
+/// If [`MakeRequestId::make_request_id`] returns `Some(_)` and the request doesn't already have a
+/// header with the same name, then the header will be inserted.
```

Comment:
Maybe worth adding the part about not overwriting the header to the module docs as well?